Add spec for Span Link and Span Continuation (#600)
elastic/apm: specs/agents/span-links.md

## Span Links

A Span or Transaction MAY link to zero or more other Spans/Transactions that are causally related.

Example use-cases for Span Links:

1. When a single transaction represents the batch processing of several messages, the agent is able to link back to the traces that have produced the messages.

> 2. When the agent receives a `traceparent` header from outside a trust boundary, it can restart the trace (creating a different trace id with its own sampling decision) and link to the originating trace.

Reviewer: ```suggestion
2. When the agent receives a `traceparent` header from outside a trust boundary, it [can restart the trace](trace_continuation.md) (creating a different trace id with its own sampling decision) and link to the originating trace.
```

## Span Links

> A Span or Transaction MAY be linked to zero or more other Spans/Transactions that are causally related.

Reviewer 2: ```suggestion
A Span or Transaction MAY link to zero or more other Spans/Transactions that are causally related.
```

> Specific use-cases for Span Links:

Reviewer 2: ```suggestion
Example use-cases for Span Links:
```

> 1. When a single transaction represents the batch processing of several messages, the agent SHOULD be able to link back to the traces that have produced the messages.

Reviewer 2: As these are examples and not actual specs for how to handle messaging, I would not use the RFC-style `SHOULD`.
Once we have a spec for how to handle span links for message tracing, we should link to that.

> 2.  When the agent receives a `traceparent` header from outside a trust boundary, it SHOULD restart the trace (creating a different trace id with its own sampling decision) and link to the originating trace.

Reviewer 2: ```suggestion
2. When the agent receives a `traceparent` header from outside a trust boundary, it SHOULD restart the trace (creating a different trace id with its own sampling decision) and link to the originating trace.
```

3. Close gap for the OTLP intake - [OTel's specification of span links](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/overview.md#links-between-spans)

> Spans and Transactions MUST collect links in the `links` array with the following fields on each item:

Reviewer 3: I wonder if the link should have type/kind to better indicate why a trace/span is linked to the current one.

Author: The labels field discussed above could potentially cover that: https://github.com/elastic/apm/pull/600#pullrequestreview-889811169

> Spans and Transactions MUST collect links in the `links` array with the following fields on each item:

Reviewer 3: Should/must agents provide some (public) api to add links to the current transaction/span?

Author: Yes, the plan is to look into extending the public API as well, but that'll be a separate spec on top of this. This one is about how agents internally handle span links.

Issue for extending the public API: https://github.com/elastic/apm/issues/594

Reviewer: @tobiasstadler I'd be happy to have opinions on the Java API for this at #594, especially because I'm not a Java guy. I'm drafting the spec there now. Briefly, I think the agent's API should be compatible with https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#specifying-links and current OTel language-specific API definitions, and in such a way that if/when we add support for span link *attributes*, they fit into the existing API smoothly.

- `trace.id`: the id of the linked trace.

> - `span.id`: the id of the linked span or transaction.

Reviewer 2: I'm good with that.
Rather than changing the name of `span.id` to something like `span_or_transaction.id`, we should rename `transaction.id` to `span.id`. Transactions are just a special type of span - a span that represents an entry into a service.
But that's a discussion for another time.

Example:

```
"links": [
  {"trace": {"id": "traceId1"}, "span": {"id": "spanId1"}},
  {"trace": {"id": "traceId2"}, "span": {"id": "spanId2"}},
]

> ```

Reviewer: Per https://github.com/elastic/apm-server/pull/7553/files#diff-7c3116c7517c49da2ce5db0a2d59625298a0c96bb28d3b9127e3a81aed59f1f1R566 it looks like these will be `trace_id` and `span_id`.

Author: Hahh, thanks, fixed.

>  "links": [
>       {"trace": {"id": "traceId1"}, "span": {"id": "spanId1"}},
>       {"trace": {"id": "traceId2"}, "span": {"id": "spanId2"}},
>     ]

Reviewer 2: ```suggestion
"links": [
  {"trace": {"id": "traceId1"}, "span": {"id": "spanId1"}},
  {"trace": {"id": "traceId2"}, "span": {"id": "spanId2"}},
]
```

> ## Trace Continuation

Reviewer 2: I'd move this section to a separate file. While it's linked to span links (pun intended), it's a different feature.

> We can expect incoming requests into an application with our agent which contains a `traceparent` header added by a 3rd party component. In this situation we end up with traces where part of the trace is outside of our system.

Reviewer 2: ```suggestion
The `traceparent` header of requests that are traces with our agents might have been added by a 3rd party component.
This situation becomes more and more common as the w3c trace context gets adopted.
If our agents blindly trust the `traceparent` header, we end up with traces where part of the trace is outside of our system and were we don't have control over the sampling rate.
```

Author: I think there are good ideas here (some I already took), 2 things I'd like to clarify:

1. You suggest 
> The `traceparent` header of requests that are traces with our agents

Are you sure about the "requests that are traces with our agent" part? I think it's hard to understand... is it maybe "requests that are traced with our agents"? 

2. 
> If our agents blindly trust the `traceparent` header, we end up with traces where part of the trace is outside of our system and were we don't have control over the sampling rate.

Do we really need to go into the sample rate here? I think currently there are some other issues when part of the trace is outside our system, so I'd not focus on specific issues. I'd just state that part of the trace is outside our system and this section of the spec intends to handle this in general.

Additionally the currently default `continue_always` strategy will allow traces where part of it is outside of our system which isn't reallyt aligned with saying "If our agents blindly trust the `traceparent` header, we end up with traces where part of the trace is outside of our system"

So overall in this part my intention was to say:
- there could be incoming requests with `tracestate` added by a 3rd party (defining what's the use case)
- this may lead to traces where part of the trace is outside our system (problem statement)
- the agent should be able to deal with this (and the way to deal with it is specified later)

I adapted this part, let me know if there is something left we should change.

Reviewer 2: sounds good, thanks

> In order to handle this properly, the agent SHOULD implement a trace continuation strategy.

Reviewer 2: ```suggestion
In order to handle this properly, the agent SHOULD offer several trace continuation strategies.
```

The agent SHOULD offer a configuration called `trace_continuation_strategy` with the following values and behavior:

- `continue_always`: The agent takes the `traceparent` header as it is and applies it to the new transaction.
- `restart_always`: The agent always creates a new trace with a new trace id. In this case the agent MUST create a link in the new transaction pointing to the original trace.

> - `restart_external`: The agent first checks the `tracestate` header. If the header contains the `es` vendor flag, it's treated as internal, otherwise (including the case when the `tracestate` header is not present) it's treated as external. In case of external calls the agent MUST create a new trace with a new trace id and MUST create a link in the new transaction pointing to the original trace.

Reviewer 2: We have also talked about being able to configure something like a project or organization id. This ID would get propagated in the tracestate header. If set, agents will check for equality of the value of the incoming tracestate header with the configured value rather than just checking if the `es` vendor key is present. The reason is that checking for the `es` key may not be enough to identify external calls as the external entity might also use Elastic APM.
I'm fine with leaving that out in the first iteration as we can always add it later on but we should be conscious around making that decision.

Author: Summerized our zoom discussion from the .NET weekly [here](https://github.com/elastic/apm/pull/600#issuecomment-1077981336) - so let's leave this out in the first iteration.

> The default is `continue_always`. 

Reviewer 2: Is this true for the .NET agent? IIRC, it does something similar to `restart_external`.

Author: > Is this true for the .NET agent? 

Yes, it is. The default is still the original behaviour: just get he value and apply it.

> IIRC, it does something similar to restart_external.

Correct, but that's not the default - [there is the `TraceContextIgnoreSampledFalse` flag](https://www.elastic.co/guide/en/apm/agent/dotnet/current/config-supportability.html#config-trace-context-ignore-sampled-false), but by default it's turned off. So the suggested default here would be fine for .NET, that's the same as the current default.

As discussed `TraceContextIgnoreSampledFalse` will be deprecated long term after this spec is implemented, and then as you say, `restart_external` will be basically the same as the deprecated flag. (Not 100% same, because in .NET we still reused the trace id, we just removed the parent id and made a new sampling decision, but I don't think this changes anything...).
